src/App.js

```diff
@@ -44,8 +90,164 @@ class App extends Component {
 
+    handleAddCategory(category) {
+        if(category.name === '') return;
+
+        const newCategory = {
+            id: this.state.nextCatId,
+            name: category.name,
+            parentId: null
+        };
+
+        let newList = this.state.categories;
+        newList.unshift(newCategory);
+
+        this.setState({
+            categories: newList,
+            nextCatId: this.state.nextCatId + 1
+        });
+    }
+
+    handleAddTask(task) {
+
+        if(task.name === '') return;
+
+        const newTask = {
+            id: this.state.nextTaskId,
+            name: task.name,
+            categoryId: task.categoryId,
+            isDone: false
+        };
+
+        let newList = this.state.tasks;
+        newList.unshift(newTask);
+
+        this.setState({
+            tasks: newList,
+            nextTaskId: this.state.nextTaskId + 1
+        });
+    }
+
+    handleAddSubcategory(parent) {
+
+        this.setState({
+            modalCategoryActive: true,
+            modalCategoryState: 'add',
+            categoryToUpdate: parent
+        });
+    }
+
+    handleUpdateCategory(node) {
+
+        this.setState({
+            modalCategoryActive: true,
+            modalCategoryState: 'update',
+            modalCategoryInputText: node.name,
+            categoryToUpdate: node
+        });
+    }
+
+    handleCancelModalCategory() {
+        this.setState({
+            modalCategoryActive: false,
+            modalCategoryState: null,
+            modalCategoryInputText: '',
+            categoryToUpdate: null
+        });
+    }
+
+    handleSubmitModalCategory(categoryName) {
+
+        if(this.state.modalCategoryState === 'update') {
+            const list1 = fromJS(this.state.categories);
+            const categoryIndex = this.state.categories.indexOf(this.state.categoryToUpdate);
+            const list2 = list1.updateIn([categoryIndex, 'name'], name => categoryName).toJS();
+
+            this.setState({
+                modalCategoryActive: false,
+                modalCategoryState: null,
+                categoryToUpdate: null,
+                categories: list2
+            });
+
+        } else if(this.state.modalCategoryState === 'add') {
+
+            if(categoryName === '') return;
+
+            const newCategory = {
+                id: this.state.nextCatId,
+                name: categoryName,
+                parentId: this.state.categoryToUpdate.id
+            };
+
+            let list2 = this.state.categories;
+            list2.unshift(newCategory);
+
+            this.setState({
+                modalCategoryActive: false,
+                modalCategoryState: null,
+                categoryToUpdate: null,
+                nextCatId: this.state.nextCatId + 1
+            });
+
+        }
+    }
+
+    handleInputChange() {
+
+        this.setState({
+            addCategoryInputText: this.categoriesComponent.addInput.value,
+            taskInputText: this.tasksComponent.taskInput.value,
```

Comment:
`this.tasksComponent` is not defined. Anyway, input should handled in the component it belongs to.